bagelbids/merge.py

```diff
@@ -122,7 +125,12 @@ def cli(bids_path, demo_path, out_path):
     demo_json = json.load(open(demo_path))
 
     context = generate_context()
-    context.update(**merge_json(bids_json, demo_json))
+
+    # TODO: revisit this implementation. It is concerningly implicit
```

Comment:
Just for my understanding, can you explain this comment further here?

Reply:
Yes, thanks for checking :). 

Essentially the question is: when and where do we add the identifier attribute and give it a value of str + UUID (I mentioned this also briefly [in the issue](https://github.com/neurobagel/bagelbids/issues/17#issuecomment-1243897104)). When we load the two `.json` files here (one for the annotated participants.tsv and the other for the parsed BIDS dataset), they don't yet come with identifiers - instead subjects are labeled with their BIDS `participant_id`, and sessions with their BIDS session_id and so on. 

If we had already added UUID identifiers here to the subjects and sessions, then merging the two json files would be weird, because the same subject (based on its label / `participant_id`) would have a different UUID identifier in the BIDS.json and the participant.json - and then we either keep one and toss the other, or make a new one. 

If we add the identifier here for the first time (which I think makes sense), then we need to somehow walk through the entire json file, get to every object (or node), and then give it an identifier attribute and a UUID string or validate the existing one against a pattern. Doing this directly in the json seems complicated, so I didn't want to do it that way. 

An alternative could have been to say: the identifier is optional, and there is some kind of method that adds an identifier / UUID after an instance has been created. In that case, we could have done:

1. Get the two json files
2. merge the json files
3. make a datamodel instance out of the merged json file
4. then walk down the model instance tree and call the `add_identifier` method for every instance, updating the overall datamodel instance as we go

I didn't want to do this either, because I couldn't find a way in pydantic to have a recursive function apply to the whole model instance - and just updating the datamodel instance bit by bit also seems complex.

So I chose this option which is:
- the identifier attribute is optional, but if no value is supplied, then a new str + UUID is autogenerated ([see here](https://github.com/neurobagel/bagelbids/pull/19/files#diff-9ac140bd6e1662b15e438c063de10913b3b0d8b9f47f1c9dd7c22e4a692e8ffdR17))
- so when I take the merged json file and make a big datamodel instance out of it, then every entity in that tree automatically gets an identifier with str + uuid

The problem with that is that it is implicit. Just from looking at the line it is not clear why I would go from JSON (dict) to datamodel instance and then back to JSON (dict). So I added the comment to explain why we do it (validation, and to add the identifier). But I still think this is not very clear, so I left the TODO flag.

All that is to say: if you have a good idea how to do this cleaner, please tell me and I'll change it.